pkg/profile/artifact.go

```diff
@@ -37,19 +34,63 @@ func (p *Profile) CreateArtifacts(ctx context.Context) error {
 		if !ok {
 			return fmt.Errorf("object %v cannot be asserted to client.Object", o)
 		}
-		p.log.Info("creating...", "kind", obj.GetObjectKind().GroupVersionKind().Kind, "resource", obj.GetName())
-		if err := p.client.Create(ctx, obj); err != nil {
-			return fmt.Errorf("failed to create %s: %w", obj.GetObjectKind().GroupVersionKind().Kind, err)
+		if err := p.reconcileArtifact(obj); err != nil {
+			return err
 		}
 	}
 
 	p.log.Info("all artifacts created")
 	return nil
 }
+func (p *Profile) reconcileArtifact(desiredObj client.Object) error {
+	existingObj := desiredObj.DeepCopyObject().(client.Object)
+	err := p.client.Get(p.ctx, client.ObjectKeyFromObject(desiredObj), existingObj)
+	if err == nil {
+		return p.updateResource(existingObj, desiredObj)
+	} else if apierrors.IsNotFound(err) {
+		p.log.Info("creating...", "kind", desiredObj.GetObjectKind().GroupVersionKind().Kind, "resource", desiredObj.GetName())
+		if err := p.client.Create(p.ctx, desiredObj); err != nil {
+			return fmt.Errorf("failed to create %s: %w", desiredObj.GetObjectKind().GroupVersionKind().Kind, err)
+		}
+		return nil
+	} else {
+		//todo update
+		return fmt.Errorf("failed to create %s: %w", desiredObj.GetObjectKind().GroupVersionKind().Kind, err)
+	}
+}
+
+func (p *Profile) updateResource(existingObj, desiredObj client.Object) error {
+	switch desiredObj := desiredObj.(type) {
+	case *sourcev1.GitRepository:
+		if !GitRepoRequiresUpdate(existingObj.(*sourcev1.GitRepository), desiredObj) {
+			return nil
+		}
+		existingObj.(*sourcev1.GitRepository).Spec = desiredObj.Spec
```

Comment:
anybody know if there is a way to not have to type converter twice :vomiting_face: ?